go/vt/vttablet/tabletserver/vstreamer/engine.go

```diff
@@ -0,0 +1,235 @@
+	vschemapb "vitess.io/vitess/go/vt/proto/vschema"
+)
+
+var (
+	once           sync.Once
+	vschemaErrors  *stats.Counter
+	vschemaUpdates *stats.Counter
+)
+
+// Engine is the engine for handling vreplication streaming requests.
+type Engine struct {
+	// cp is initialized by InitDBConfig
+	cp *mysql.ConnParams
+
+	// mu protects isOpen, streamers, streamIdx and kschema.
+	mu sync.Mutex
+
+	isOpen bool
+	// wg is incremented for every Stream, and decremented on end.
+	// Close waits for all current streams to end by waiting on wg.
+	wg        sync.WaitGroup
+	streamers map[int]*vstreamer
+	streamIdx int
+
+	// watcherOnce is used for initializing kschema
+	// and setting up the vschema watch. It's guaranteed that
+	// no stream will start until kschema is initialized by
+	// the first call through watcherOnce.
+	watcherOnce sync.Once
+	kschema     *vindexes.KeyspaceSchema
+
+	// The following members are initialized once at the beginning.
+	ts       srvtopo.Server
+	se       *schema.Engine
+	keyspace string
+	cell     string
+}
+
+// NewEngine creates a new Engine.
+// Initialization sequence is: NewEngine->InitDBConfig->Open.
+// Open and Close can be called multiple times and are idempotent.
+func NewEngine(ts srvtopo.Server, se *schema.Engine) *Engine {
+	vse := &Engine{
+		streamers: make(map[int]*vstreamer),
+		kschema:   &vindexes.KeyspaceSchema{},
+		ts:        ts,
+		se:        se,
+	}
+	once.Do(func() {
+		vschemaErrors = stats.NewCounter("VSchemaErrors", "Count of VSchema errors")
+		vschemaUpdates = stats.NewCounter("VSchemaUpdates", "Count of VSchema updates. Does not include errors")
+		http.Handle("/debug/tablet_vschema", vse)
+	})
+	return vse
+}
+
+// InitDBConfig performs saves the required info from dbconfigs for future use.
+func (vse *Engine) InitDBConfig(dbcfgs *dbconfigs.DBConfigs) {
+	vse.cp = dbcfgs.DbaWithDB()
+}
+
+// Open starts the Engine service.
+func (vse *Engine) Open(keyspace, cell string) error {
+	vse.mu.Lock()
+	defer vse.mu.Unlock()
+	if vse.isOpen {
+		return nil
+	}
+	vse.isOpen = true
+	vse.keyspace = keyspace
+	vse.cell = cell
+	return nil
+}
+
+// Close closes the Engine service.
+func (vse *Engine) Close() {
+	func() {
+		vse.mu.Lock()
+		defer vse.mu.Unlock()
+		if !vse.isOpen {
+			return
+		}
+		for _, s := range vse.streamers {
+			// cancel is non-blocking.
+			s.Cancel()
+		}
+		vse.isOpen = false
+	}()
+
+	// Wait only after releasing the lock because the end of every
+	// stream will use the lock to remove the entry from streamers.
+	vse.wg.Wait()
+}
+
+func (vse *Engine) vschema() *vindexes.KeyspaceSchema {
+	vse.mu.Lock()
+	defer vse.mu.Unlock()
+	return vse.kschema
+}
+
+// Stream starts a new stream.
+func (vse *Engine) Stream(ctx context.Context, startPos mysql.Position, filter *binlogdatapb.Filter, send func([]*binlogdatapb.VEvent) error) error {
+	// Ensure kschema is initialized and the watcher is started.
+	vse.watcherOnce.Do(vse.setWatch)
```

Comment:
Instead of doing a `watcherOnce` here, I would initialize this as part of the `Open` or even when creating the engine. That would be consistent on how we do it in vschema_manager.